### ISSUE TITLE
FilePicker.getFile(type: FileType.ANY); does not work

### ISSUE THREAD

**Issue by gjrjf** [miguelpruivo/flutter_file_picker#195]
**Issue details**
1. Platform (Android)
2. OS version
3. File type (filter)
4. Steps to reproduce

**Error Log**
[ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: MissingPluginException(No implementation found for method ANY on channel file_picker)
E/flutter (21367): #0      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:319:7)
E/flutter (21367): <asynchronous suspension>
E/flutter (21367): #1      FilePicker._getPath (package:file_picker/file_picker.dart:65:39)
E/flutter (21367): #2      FilePicker.getFile (package:file_picker/file_picker.dart:46:15)
E/flutter (21367): #3      ChatScreenState.uploadContents (package:Hwa/pages/chatting/chatroom_page.dart:370:36)
E/flutter (21367): #4      ChatScreenState.chatIconOpen.<anonymous closure> (package:Hwa/pages/chatting/chatroom_page.dart:1009:30)
E/flutter (21367): #5      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
E/flutter (21367): #6      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:486:11)
E/flutter (21367): #7      BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:264:5)
E/flutter (21367): #8      BaseTapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:236:7)
E/flutter (21367): #9      GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:156:27)
E/flutter (21367): #10     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:222:20)
E/flutter (21367): #11     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:22)
E/flutter (21367): #12     GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:156:7)
E/flutter (21367): #13     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:102:7)
E/flutter (21367): #14     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:86:7)
E/flutter (21367): #15     _rootRunUnary (dart:async/zone.dart:1138:13)
E/flutter (21367): #16     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
E/flutter (21367): #17     _CustomZone.runUnaryGuarded (dart:async/zone.dart:933:7)
E/flutter (21367): #18     _invoke1 (dart:ui/hooks.dart:273:10)
E/flutter (21367): #19     _dispatchPointerDataPacket (dart:ui/hooks.dart:182:5)

**Flutter Version details**
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale ko-KR)
    • Flutter version 1.12.13+hotfix.5 at /Users/hk/Documents/dev/flutter
    • Framework revision 27321ebbad (4 weeks ago), 2019-12-10 18:15:01 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/hk/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/hk/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2017.1.6)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] Connected device (1 available)
    • SM N971N • R3CMB0EE6XZ • android-arm64 • Android 9 (API 28)

**Additional context**

**Comment by miguelpruivo** [comment]
@gjrjf whats the issue? What are you trying to achieve?

**Comment by gjrjf** [comment]
> @gjrjf whats the issue? What are you trying to achieve?

at  " ^1.4.3+2 " Version
FilePicker.getFile(type: FileType.ANY);  => does not work

errLog : Unhandled Exception: MissingPluginException(No implementation found for method ANY on channel file_picker)

**Comment by miguelpruivo** [comment]
Hum, there must be something missing. If it would not work, it would affect several more Android users for sure. 

Which version of Flutter are you using? Can you try running the example app provided with the plugin and see if it works for you?

Thank you.

**Comment by gjrjf** [comment]
> Hum, there must be something missing. If it would not work, it would affect several more Android users for sure.
> 
> Which version of Flutter are you using? Can you try running the example app provided with the plugin and see if it works for you?
> 
> Thank you.

Flutter Version is

flutter --version
Flutter 1.12.13+hotfix.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 27321ebbad (4 weeks ago) • 2019-12-10 18:15:01 -0800
Engine • revision 2994f7e1e6
Tools • Dart 2.7.0

Thanks

**Comment by Boehrsi** [comment]
I can confirm this problem. It's related to https://github.com/miguelpruivo/flutter_file_picker/issues/196.
As soon as an App uses the Flutter 1.12.x + Android v2 embedding the plugin stops functioning. The reason is https://github.com/miguelpruivo/flutter_file_picker/blob/master/android/src/main/java/com/mr/flutter/plugin/filepicker/FilePickerPlugin.java#L43, as the registrar will always return `null`for the `activity()` call.

With Android v2 embedding the registrar isn't used anymore and instead the ActivityPluginBinding interface should get used https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/engine/plugins/activity/ActivityPluginBinding.java.

A complete guide can be found here https://flutter.dev/docs/development/packages-and-plugins/plugin-api-migration.

**Comment by Boehrsi** [comment]
As this is currently blocking me I would just try to prepare a PR today, would that be okay for you @miguelpruivo or would you like to do this yourself?

**Comment by miguelpruivo** [comment]
@Boehrsi no problem, since this is blocking, I'll take a look even today and probably deploy a new version with it on pub.dev. I'll keep you updated.

Thank you!

**Comment by Boehrsi** [comment]
@miguelpruivo thanks a lot :+1: . As I had some spare time I already did some changes in my fork (https://github.com/Boehrsi/plugins_flutter_file_picker/commits/192_embeddingV2) which already works now. If you like I could create a PR or you can just have a look, just tell me if I should create the PR :). Haven't had the time to clean up the commit until now (formatting and stuff).

**Comment by kxait** [comment]
Any way to work around this quickly?

**Comment by Boehrsi** [comment]
@kxait you could do something like
```
file_picker:
    git:
      url: git://github.com/Boehrsi/plugins_flutter_file_picker
      ref: 192_embeddingV2
```
in your `pubspec.yaml`. It's my fork which works fine (at least for me no bugs occurred). If you don't like to depend on someone else repository directly (as it could change without your knowledge at any point in time) you can just copy / fork my fork and depend on this.
But I would assume this is fixed in the near future with an official release :).

**Comment by kxait** [comment]
> @kxait you could do something like
> 
> ```
> file_picker:
>     git:
>       url: git://github.com/Boehrsi/plugins_flutter_file_picker
>       ref: 192_embeddingV2
> ```
> 
> in your `pubspec.yaml`. It's my fork which works fine (at least for me no bugs occurred). If you don't like to depend on someone else repository directly (as it could change without your knowledge at any point in time) you can just copy / fork my fork and depend on this.
> But I would assume this is fixed in the near future with an official release :).

Thanks so much, I needed a quick fix and didn't really know how to do this. Hoping on the update soon!

**Comment by Boehrsi** [comment]
@kxait no problem :), but I would generally not push releases with dependencies like this. For testing and unblocking development it's nice, but as I'm not the maintainer of the project and don't know all internal flows of the plugin, my fix could be not as stable as this plugin normally is. Please be aware.

**Comment by kxait** [comment]
> @kxait no problem :), but I would generally not push releases with dependencies like this. For testing and unblocking development it's nice, but as I'm not the maintainer of the project and don't know all internal flows of the plugin, my fix could be not as stable as this plugin normally is. Please be aware.

Thanks for letting me know, but I'm just working on a hobby project :) 
By the way, is an issue with flutter unexpectedly dropping and quitting after selecting a file in any way related to this?

**Comment by Boehrsi** [comment]
Hmm, I didn't had something similar, so can't tell sorry. Yeah, for a hobby project I would assume that temp solution should be fine :).

**Comment by miguelpruivo** [comment]
I’ll be working on this during the weekend so you can expect a pubspec release until Sunday. 

Sorry guys, time has been short.

**Comment by kxait** [comment]
@Boehrsi has this been working for you? I can't get around the issue where the app crashes after exiting out of the file picker (regardless if file was chosen or not), reproduced this in the example app and on different devices (different AVDs too). Keep in mind this is on your fork, not the official repo so I'm not going to submit an issue.

**Comment by Boehrsi** [comment]
@kxait sorry for the problems, I've just tested it again in my project without problems. If you checkout my fork (it contains an adjusted example app), does this work for you?

**Comment by kxait** [comment]
@Boehrsi some progress: I modified my AndroidManifest.xml in line with the one with your fork, the changes:

- removed application android:name
- removed activity android:name and android:launchMode
- added activity android:name
- modified activity meta-data android:name and android:value

The file picker no longer crashes the app upon exiting it, but now my DefaultCacheManager is throwing MissingPluginException. I will continue to investigate.

**Comment by kxait** [comment]
Figured it out. Activity parameter android:name can be either `".MainActivity"` or `"io.flutter.embedding.android.FlutterActivity"` for the app to launch, the cache works with `.MainActivity` and the picker makes the app crash, but `io.flutter.embedding.android.FlutterActivity` causes the app to crash with:
`MissingPluginException(No implementation found for method getDatabasesPath on channel com.tekartik.sqflite)` and
`MissingPluginException(No implementation found for method getTemporaryDirectory on channel plugins.flutter.io/path_provider)`, 
but the file picker works. Any idea how to get around this? I'm pretty sure this is a problem with embedding v2 compatiblity across the modules, (namely sqflite?) but no idea how to fix this right now.

**Comment by Boehrsi** [comment]
It seems like the other mentioned plugins aren't ready for the v2 embedding yet. In general this means only v1 embedding is possible for you. As plugins using v2 *should be backward compatible*, but if a plugin only is v1 compatible, v2 can't be used. 
The strange thing are the problems you having with my fork, as I tested it with the v1 and v2 embedding, both without problems, so it should work.

**Comment by miguelpruivo** [comment]
Guys, just to let you know that I didn't forget this and I was hoping to have some time by the weekend but it turns out that I didn't. 

I'm aware that this is a high priority issue that might be affecting a few users and I'll try to solve it even before Wednesday (maybe tomorrow) because otherwise only by the next week (due to Flutter Europe event).

I'll keep you updated. 

Thank you!

**Comment by miguelpruivo** [comment]
@Boehrsi I've spent some time around this and it looks like it will take a bit more time that I expected in order to properly implement it. Just need a few more details from you, since you have further worked into it.

From what I saw, `FlutterPluginBinding`  replaces `Registrar` in V2, but does that mean that V2 is _only_ used in Android apps that are trying to embed Flutter in? What were you trying to do when you first experienced the issue?

Thanks!

**Comment by Boehrsi** [comment]
> `FlutterPluginBinding` replaces `Registrar`

@miguelpruivo sure no problem :).

- `FlutterPluginBinding` replaces `Registrar` is true for user who embed via V2, if v1 is used the Registrar part should stay. That's the reason why I kept the `registerWith` class in my PR (https://github.com/Boehrsi/plugins_flutter_file_picker/commit/64596b1d117757d63e86fc8141cdf9062399bfa6#diff-8f1f188de7002bd224459770dd4781f2L41) it should provide backward compatibility. For more information please see https://flutter.dev/docs/development/packages-and-plugins/plugin-api-migration#upgrade-steps
- V2 is only for Android, absolutely correct, no iOS changes should be required
- I encountered the error the first time after updating my Flutter to 1.12.x + I switched to embedding V2 for some complex background plugin stuff I need for an app. After the this error occurred first, as soon as I tried to make any method call against the plugin.

**Comment by rshrc** [comment]
@miguelpruivo any updates on this issue?

**Comment by miguelpruivo** [comment]
@rshrc I started a draft on separate branch working on this, but haven't had the time yet to get it done. There are multiple things to handle that can't be missed.

Btw, can one confirm is this affects _only_ those that are interested into embed Flutter modules in Android apps?

cc: @Boehrsi

**Comment by Boehrsi** [comment]
Sorry for the late response. I'm facing this issue in a pure Flutter app, with some native components though.

**Comment by miguelpruivo** [comment]
Just FYI I have plans for releasing V2 support with some improved Android implementation in the next days.

I’ll keep you updated.

**Comment by miguelpruivo** [comment]
V2 embedding support was added and published with [1.5.0](https://pub.dev/packages/file_picker#150).

Have in mind, that along with it, now `FileType` has been refactored to match camelCase enum Dart guideline, so you'll need to refactor your code to match accordingly (eg. `FileType.ALL` now is `FileType.all`).

Thank you all for being patient. 🎉
